modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -109,23 +112,34 @@ public Netty4HttpPipeliningHandler(
     public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
         activityTracker.startActivity();
         try {
-            assert msg instanceof FullHttpRequest : "Should have fully aggregated message already but saw [" + msg + "]";
-            final FullHttpRequest fullHttpRequest = (FullHttpRequest) msg;
-            final Netty4HttpRequest netty4HttpRequest;
-            if (fullHttpRequest.decoderResult().isFailure()) {
-                final Throwable cause = fullHttpRequest.decoderResult().cause();
-                final Exception nonError;
-                if (cause instanceof Error) {
-                    ExceptionsHelper.maybeDieOnAnotherThread(cause);
-                    nonError = new Exception(cause);
+            assert msg instanceof PipelinedHttpObject : "Should have pipelined message already but saw [" + msg + "]";
+            final var sequence = ((PipelinedHttpObject) msg).sequence();
+            if (msg instanceof HttpRequest request) {
+                final Netty4HttpRequest netty4HttpRequest;
+                if (request.decoderResult().isFailure()) {
+                    final Throwable cause = request.decoderResult().cause();
+                    final Exception nonError;
+                    if (cause instanceof Error) {
+                        ExceptionsHelper.maybeDieOnAnotherThread(cause);
+                        nonError = new Exception(cause);
+                    } else {
+                        nonError = (Exception) cause;
+                    }
+                    netty4HttpRequest = new Netty4HttpRequest(sequence, (FullHttpRequest) request, nonError);
                 } else {
-                    nonError = (Exception) cause;
+                    if (request instanceof FullHttpRequest fullRequest) {
+                        netty4HttpRequest = new Netty4HttpRequest(sequence, fullRequest);
+                    } else {
+                        var contentPublisher = new Netty4RequestContentPublisher(ctx.channel());
+                        netty4HttpRequest = new Netty4HttpRequest(sequence, request, contentPublisher);
+                        currentRequest = netty4HttpRequest;
```

Comment:
note: I attach current request to Handler object and following chunks will go to the same object until we see next request